optuna/storages/_rdb/storage.py

```diff
@@ -802,28 +811,21 @@ def _get_trials(
                 assert isinstance(states, Iterable)
                 query = query.filter(models.TrialModel.state.in_(states))
 
-            trial_ids = query.all()
-
-            trial_ids = set(
-                trial_id_tuple[0]
-                for trial_id_tuple in trial_ids
-                if trial_id_tuple[0] not in excluded_trial_ids
-            )
             try:
-                trial_models = (
-                    session.query(models.TrialModel)
-                    .options(sqlalchemy_orm.selectinload(models.TrialModel.params))
-                    .options(sqlalchemy_orm.selectinload(models.TrialModel.values))
-                    .options(sqlalchemy_orm.selectinload(models.TrialModel.user_attributes))
-                    .options(sqlalchemy_orm.selectinload(models.TrialModel.system_attributes))
-                    .options(sqlalchemy_orm.selectinload(models.TrialModel.intermediate_values))
-                    .filter(
-                        models.TrialModel.trial_id.in_(trial_ids),
-                        models.TrialModel.study_id == study_id,
+                if len(included_trial_ids) > 0 and trial_id_greater_than > -1:
+                    _query = query.filter(
+                        sqlalchemy.or_(
+                            models.TrialModel.trial_id.in_(included_trial_ids),
+                            models.TrialModel.trial_id > trial_id_greater_than,
+                        )
                     )
-                    .order_by(models.TrialModel.trial_id)
-                    .all()
-                )
+                elif trial_id_greater_than > -1:
+                    _query = query.filter(models.TrialModel.trial_id > trial_id_greater_than)
```

Comment:
Can we reuse the `query` variable and remove the `else` block?

```suggestion
                    query = query.filter(models.TrialModel.trial_id > trial_id_greater_than)
```

Reply:
The variable `query` is also used in the `except` block in which we don't add the filtering to the SQL query, so I use another name of variable here.

Reply:
I see. I noticed another issue and have one concern:

1. Currently, `selectinload` is used even in the `except` block, which is likely unintentional.
2. Although this isn’t related to your changes, the logic in the `except` block isn't tested at all.

What are your thoughts on these?

Reply:
Thanks for the comments.

> Currently, selectinload is used even in the except block, which is likely unintentional.

I think we have already used `selectinload` in this except block ([ref](https://github.com/optuna/optuna/blob/master/optuna/storages/_rdb/storage.py#L865)). Do you think we should avoid this? The processing enters the except block due to the specification of sqlite that the number of maximum allowed variables using IN is 999 as noted in the above comments in codes and the document, so I think it is reasonable to use `selectinload` here as we do in the other cases.

> Although this isn’t related to your changes, the logic in the except block isn't tested at all.

In my understanding, this except block is already tested [here](https://github.com/optuna/optuna/blob/master/tests/storages_tests/test_storages.py#L963). However, this test case is not obvious to understand what it tests.

Reply:
> I think we have already used selectinload in this except block (ref).

Ah, you are absolutely correct!

> In my understanding, this except block is already tested here. However, this test case is not obvious to understand what it tests.

As for unit tests, it seems that the except block isn’t actually tested in the master branch.

```
$ git co master
$ python3
>>> import optuna
>>> storage = optuna.storages.RDBStorage("sqlite:///test.db")
>>> study_id = storage.create_new_study(directions=[optuna.study.StudyDirection.MINIMIZE])
[I 2024-10-29 18:29:53,124] A new study created in RDB with name: no-name-56f4ff56-e9e8-488a-8c93-d88e312ad97b
>>> storage.create_new_trial(study_id)
1
>>> trials = storage._get_trials(study_id, states=None, excluded_trial_ids=set(range(500000)))
>>>
```

That said, this PR also addresses the test case, so my concern is effectively resolved. Thank you for your great work!

```
$ gh pr checkout 5704
$ python3
>>> import optuna
>>> storage = optuna.storages.RDBStorage("sqlite:///test.db")
>>> study_id = storage.create_new_study(directions=[optuna.study.StudyDirection.MINIMIZE])
[I 2024-10-29 18:29:53,124] A new study created in RDB with name: no-name-56f4ff56-e9e8-488a-8c93-d88e312ad97b
>>> trial_id = storage.create_new_trial(study_id)
>>> trial_id_greater_than = trial_id + 500000
>>> trials = storage._get_trials(
...            study_id,
...            states=None,
...            included_trial_ids=set(range(500000)),
...            trial_id_greater_than=trial_id_greater_than,
...        )
:
(Background on this error at: https://sqlalche.me/e/20/e3q8). Falling back to a slower alternative.
```